# 1. Loading the data

In [ ]:
import pandas as pd

In [ ]:
# Initially loading 1000 rows of each dataset to understand the structure and content
summary_df = pd.read_csv('job_summary.csv', nrows=1000)
skills_df = pd.read_csv('job_skills.csv', nrows=1000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=1000)


# 2. Exploring the data

## Looking at the first few rows

In [ ]:
print(f"First 5 rows of job_summary.csv:\n{summary_df.head()}\n\n\n")
print(f"First 5 rows of job_skills.csv:\n{skills_df.head()}\n\n\n")
print(f"First 5 rows of linkedin_job_postings.csv:\n{postings_df.head()}\n\n\n")


First 5 rows of job_summary.csv:
                                            job_link  \
0  https://www.linkedin.com/jobs/view/restaurant-...   
1  https://www.linkedin.com/jobs/view/med-surg-re...   
2  https://www.linkedin.com/jobs/view/registered-...   
3  https://uk.linkedin.com/jobs/view/commercial-a...   
4  https://www.linkedin.com/jobs/view/store-manag...   

                                         job_summary  
0  Rock N Roll Sushi is hiring a Restaurant Manag...  
1  Schedule\n: PRN is required minimum 12 hours p...  
2  Description\nIntroduction\nAre you looking for...  
3  Commercial account executive\nSheffield\nFull ...  
4  Address:\nUSA-CT-Newington-44 Fenn Road\nStore...  



First 5 rows of job_skills.csv:
                                            job_link  \
0  https://www.linkedin.com/jobs/view/housekeeper...   
1  https://www.linkedin.com/jobs/view/assistant-g...   
2  https://www.linkedin.com/jobs/view/school-base...   
3  https://www.linkedin.com/jobs/view/ele

## Count of dimensions

In [ ]:
print(f"Shape of job_summary.csv: {summary_df.shape}")
print(f"Shape of job_skills.csv: {skills_df.shape}")
print(f"Shape of linkedin_job_postings.csv: {postings_df.shape}")

Shape of job_summary.csv: (1000, 2)
Shape of job_skills.csv: (1000, 2)
Shape of linkedin_job_postings.csv: (1000, 14)


## Column names

In [ ]:
print(f"Column names in job_summay.csv: {summary_df.columns.tolist()}")
print(f"Column names in job_skills.csv: {skills_df.columns.tolist()}")
print(f"Column names in linkedin_job_postings.csv: {postings_df.columns.tolist()}")

Column names in job_summay.csv: ['job_link', 'job_summary']
Column names in job_skills.csv: ['job_link', 'job_skills']
Column names in linkedin_job_postings.csv: ['job_link', 'last_processed_time', 'got_summary', 'got_ner', 'is_being_worked', 'job_title', 'company', 'job_location', 'first_seen', 'search_city', 'search_country', 'search_position', 'job_level', 'job_type']


## Datatypes of each column

In [ ]:
print(f"Job Summary columns datatypes:\n{summary_df.dtypes}\n{"-"*50}\n")
print(f"Job Skills columns datatypes:\n{skills_df.dtypes}\n{"-"*50}\n")
print(f"LinkedIn Job Postings columns datatypes:\n{postings_df.dtypes}\n{"-"*50}\n")

Job Summary columns datatypes:
job_link       str
job_summary    str
dtype: object
--------------------------------------------------

Job Skills columns datatypes:
job_link      str
job_skills    str
dtype: object
--------------------------------------------------

LinkedIn Job Postings columns datatypes:
job_link               str
last_processed_time    str
got_summary            str
got_ner                str
is_being_worked        str
job_title              str
company                str
job_location           str
first_seen             str
search_city            str
search_country         str
search_position        str
job_level              str
job_type               str
dtype: object
--------------------------------------------------



## Checking for missing values (First 1000 rows)

In [ ]:
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")

Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link      0
job_skills    0
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link               0
last_processed_time    0
got_summary            0
got_ner                0
is_being_worked        0
job_title              0
company                0
job_location           0
first_seen             0
search_city            0
search_country         0
search_position        0
job_level              0
job_type               0
dtype: int64
 --------------------------------------------------



## Checking for missing values (First 100000 rows)

In [ ]:
summary_df = pd.read_csv('job_summary.csv', nrows=100000)
skills_df = pd.read_csv('job_skills.csv', nrows=100000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=100000)


In [ ]:
print("Missing value count across 100000 records")
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")

Missing value count across 100000 records
Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link       0
job_skills    66
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link               0
last_processed_time    0
got_summary            0
got_ner                0
is_being_worked        0
job_title              0
company                1
job_location           0
first_seen             0
search_city            0
search_country         0
search_position        0
job_level              0
job_type               0
dtype: int64
 --------------------------------------------------



## Checking for missing values (First 200000 rows)

In [ ]:
summary_df = pd.read_csv('job_summary.csv', nrows=200000)
skills_df = pd.read_csv('job_skills.csv', nrows=200000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=200000)

print("Missing value count across 100000 records")
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")


Missing value count across 100000 records
Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link        0
job_skills    112
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link               0
last_processed_time    0
got_summary            0
got_ner                0
is_being_worked        0
job_title              0
company                2
job_location           3
first_seen             0
search_city            0
search_country         0
search_position        0
job_level              0
job_type               0
dtype: int64
 --------------------------------------------------



## Checking for missing values (First 500000 rows)

In [ ]:
summary_df = pd.read_csv('job_summary.csv', nrows=500000)
skills_df = pd.read_csv('job_skills.csv', nrows=500000)
postings_df = pd.read_csv('linkedin_job_postings.csv', nrows=500000)

print("Missing value count across 100000 records")
print(f"Job Summary columns null count: \n{summary_df.isnull().sum()}\n {"-"*50}\n")
print(f"Job Skills columns null count: \n{skills_df.isnull().sum()}\n {"-"*50}\n")
print(f"LinkedIn Job Postings columns null count: \n{postings_df.isnull().sum()}\n {"-"*50}\n")

Missing value count across 100000 records
Job Summary columns null count: 
job_link       0
job_summary    0
dtype: int64
 --------------------------------------------------

Job Skills columns null count: 
job_link        0
job_skills    276
dtype: int64
 --------------------------------------------------

LinkedIn Job Postings columns null count: 
job_link                0
last_processed_time     0
got_summary             0
got_ner                 0
is_being_worked         0
job_title               0
company                 3
job_location           10
first_seen              0
search_city             0
search_country          0
search_position         0
job_level               0
job_type                0
dtype: int64
 --------------------------------------------------



# 3. Cleaning the data

## Dropping columns with NULL skills

In [ ]:
skills_df = skills_df.dropna(subset=['job_skills'])

print(f"Missing values in job skills dataset after dropping null job skills:\n {skills_df.isnull().sum()}")

Missing values in job skills dataset after dropping null job skills:
 job_link      0
job_skills    0
dtype: int64


## Fixing Iconsistent Text
- Lowercase all records
- Stripping whitespaces

In [ ]:
# Job Titles
postings_df['job_title'] = postings_df['job_title'].str.lower().str.strip()
# Job Skills
skills_df['job_skills'] = skills_df['job_skills'].str.lower().str.strip()

# 4. Reducing/Filtering the data

## Numerosity Reduction

### Unique job titles and their counts

In [ ]:
print(f"Unique job titles in linkedin job postings and their counts:\n {postings_df['job_title'].value_counts()}\n\n")

Unique job titles in linkedin job postings and their counts:
 job_title
customer service representative                                                                                                        3582
lead sales associate-ft                                                                                                                2963
store manager                                                                                                                          2523
assistant manager                                                                                                                      2195
first year tax professional                                                                                                            2115
                                                                                                                                       ... 
cub foods quarry - part time cake decorator - progressive scale from $15.19 to $17.50 / 

The dataset contains 250,825 unique job titles across all industries. However, for our project we are only interested in tech related job titles with sufficient postings to reliably analyze skill requirements.

### Filtering out all non-technical job titles

#### Creating list of technical job titles

In [ ]:
tech_keywords = [
    # Software
    'software engineer', 'software developer', 'software architect', 'software designer',

    # Web
    'frontend', 'front-end', 'backend', 'back-end', 'full stack', 'fullstack', 'web developer', 'web engineer', 'web designer',

    # Data
    'data scientist', 'data analyst', 'data engineer', 'data architect', 'data visualization', 'data science', 'data analytics',

    # ML/AI
    'machine learning', 'deep learning', 'artificial intelligence', 'ai engineer', 'ml engineer', 'nlp', 'natural language processing', 'computer vision',

    # DevOps/Cloud
    'devops', 'cloud engineer', 'site reliability', 'sre', 'platform engineer', 'aws', 'azure', 'gcp', 'google cloud', 'amazon web services',

    # Mobile
    'android developer', 'ios developer', 'mobile developer', 'mobile engineer', 'mobile app developer', 'mobile application developer',

    # Security
    'cybersecurity', 'security engineer', 'penetration tester', 'security analyst', 'information security', 'infosec',

    # Database
    'database administrator', 'dba', 'database engineer', 'database developer', 'sql developer', 'sql engineer',

    # General
    'programmer', 'software qa', 'quality assurance engineer', 'test engineer', 'automation engineer', 'technical lead', 'tech lead', 'technology lead', 'cto', 'chief technology officer',

    # Common abbreviations safe to use
    'ui/ux', 'ux designer', 'ui designer',
    'it support', 'it specialist',
    'network engineer', 'blockchain'
]

#### Checking count of technical jobs in 500000 job postings

In [ ]:
tech_count = postings_df['job_title'].str.contains(
    '|'.join(tech_keywords), case=False, na=False
).sum()

print(f"Tech related job postings: {tech_count}")
print(f"Out of total: {len(postings_df)}")

Tech related job postings: 26683
Out of total: 500000


We have 26683 technical jobs in total 500000 jobs

#### Checking count of technical jobs in the entire dataset (1.3 Million job postings)

In [ ]:
postings_df = pd.read_csv('linkedin_job_postings.csv')

In [ ]:
tech_count = postings_df['job_title'].str.contains(
    '|'.join(tech_keywords), case=False, na=False
).sum()

print(f"Tech related job postings: {tech_count}")
print(f"Out of total: {len(postings_df)}")

Tech related job postings: 72385
Out of total: 1348454


So, in the total dataset, we have 72385 technical jobs in total 1.3 million job postings

#### Keeping only technical jobs and filtering out the rest

In [ ]:
postings_df = postings_df[postings_df['job_title'].str.contains(
    '|'.join(tech_keywords), case=False, na=False)]

#### Shape after removing non-technical jobs

In [ ]:
postings_df.shape

(72385, 14)

#### Setting minimum count threshold (At least 100 job postings for each job title) and filtering further

In [ ]:
title_counts = postings_df['job_title'].value_counts()
valid_titles = title_counts[title_counts > 100].index
postings_df = postings_df[postings_df['job_title'].isin(valid_titles)]

#### Shape after further filtering rare job titles

In [ ]:
postings_df.shape

(11432, 2)

### Dimensionality Reduction

We are only interested in job_titles, hence, we remove all other attibutes

In [ ]:
postings_df = postings_df[['job_link' , 'job_title']]

print(f"Shape after filtering and selecting columns: {postings_df.shape}")

Shape after filtering and selecting columns: (11432, 2)


#### Job title counts in reduced postings dataset

In [ ]:
print(f"Unique job titles in linkedin job postings and their counts:\n {postings_df['job_title'].value_counts()}\n\n")

Unique job titles in linkedin job postings and their counts:
 job_title
Senior Software Engineer                                             962
Quality Inspector                                                    898
Conductor:Entrega a Domicilio                                        810
Conductor:Entrega a Domicilio - MT                                   648
Senior Network Engineer                                              586
Network Engineer                                                     510
Quality Control Inspector                                            359
Construction Inspector                                               329
Senior Embedded Software Engineer                                    323
Software Engineer                                                    320
Senior Data Engineer                                                 289
Network Automation Engineer                                          253
Senior DDI Network Engineer                         

As the above list still has a lot of false positives (non-technical jobs), we will further filter them out.

In [ ]:
exclude_keywords = [
    'inspector', 'instructor', 'conductor', 'veterinarian',
    'entrega', 'domicilio', 'geotechnical', 'construction',
    'building inspector', 'fitness', 'exercise', 'swim',
    'assistant director', 'information security office consultant'
]
postings_df = postings_df[~postings_df['job_title'].str.contains(
    '|'.join(exclude_keywords), case=False, na=False
)]

Shape after further filtering out false positives

In [ ]:
postings_df.shape

(6874, 2)

#### Job title counts in postings dataset after removing false positives

In [ ]:
print(f"Unique job titles in linkedin job postings and their counts:\n {postings_df['job_title'].value_counts()}\n\n")

Unique job titles in linkedin job postings and their counts:
 job_title
Senior Software Engineer                                       962
Senior Network Engineer                                        586
Network Engineer                                               510
Senior Embedded Software Engineer                              323
Software Engineer                                              320
Senior Data Engineer                                           289
Network Automation Engineer                                    253
Senior DDI Network Engineer                                    223
Senior Backend Engineer                                        204
Lead Software Engineer                                         192
Senior Software Developer                                      191
Senior Network Engineer, Load Balancer Specialist              172
Senior Data Analyst                                            165
Data Engineer                                            

# 5. Data Integration

We are only interested in job titles and job skills, so we will be merging skills_df and postings_df on "job_links". We don't need job_summary.csv in our case.

In [ ]:
skills_df = pd.read_csv('job_skills.csv')

## Merging Job_Skills.csv and Job_Postings.csv

In [ ]:
merged_job_df = pd.merge(postings_df, skills_df, on='job_link', how='inner')

### Removing "job_link" in merged dataset

In [ ]:
merged_job_df = merged_job_df.drop(columns=['job_link'])

### Exploring merged dataset

In [ ]:
merged_job_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6271 entries, 0 to 6270
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   job_title   6271 non-null   str  
 1   job_skills  6270 non-null   str  
dtypes: str(2)
memory usage: 98.1 KB


There seems to be a record with null skills, removing it.

In [ ]:
merged_job_df = merged_job_df.dropna(subset=['job_skills'])

In [ ]:
merged_job_df.info()

<class 'pandas.DataFrame'>
Index: 6270 entries, 0 to 6270
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   job_title   6270 non-null   str  
 1   job_skills  6270 non-null   str  
dtypes: str(2)
memory usage: 147.0 KB


# 6. Data Transformation

Jobs of different fields are spread across many job titles, so, we will be categorizing them as:
1. Software Engineering
- Senior/Lead/Staff Software Engineer (all variants)
- Software Engineer III
- Sr. Software Engineer
- Senior Software Developer

2. Backend Engineering
- Senior Backend Engineer (all variants)

3. Network Engineering
- Senior/Network Engineer (all variants)
- Senior DDI Network Engineer
- Network Automation Engineer
- Senior Network Engineer Load Balancer Specialist
- Network Operations Technical Lead

4. Data Engineering
- Senior/Lead Data Engineer
- Data Engineer

5. Data Science & Analytics
- Data Scientist
- Senior Data Scientist
- Data Analyst
- Senior Data Analyst
- Staff Machine Learning Engineer
- Data Architect

6. DevOps & Automation
- Senior DevOps Engineer
- Automation Engineer
- Senior Test Engineer

7. IT & Security
- IT Specialist
- Information Technology (IT) Specialist (Security - INFOSEC)

## Categorizing Job titles

In [ ]:
def categorize_job_title(title):
    title = title.lower()
    if 'backend engineer' in title:
        return 'Backend Engineering'
    elif any(x in title for x in ['software engineer', 'software developer']):
        return 'Software Engineering'
    elif any(x in title for x in ['network engineer', 'network automation', 'ddi', 'network operations']):
        return 'Network Engineering'
    elif 'data engineer' in title:
        return 'Data Engineering'
    elif any(x in title for x in ['data scientist', 'data analyst', 'machine learning', 'data architect']):
        return 'Data Science & Analytics'
    elif any(x in title for x in ['devops', 'automation engineer', 'test engineer']):
        return 'DevOps & Automation'
    elif any(x in title for x in ['it specialist', 'information technology', 'infosec']):
        return 'IT & Security'
    else:
        return 'Other'

In [ ]:
merged_job_df['job_title'] = merged_job_df['job_title'].apply(categorize_job_title)

### Job title counts in merged dataset after categorization

In [ ]:
merged_job_df['job_title'].value_counts()

job_title
Software Engineering        2621
Network Engineering         1467
Data Science & Analytics     759
Data Engineering             557
DevOps & Automation          322
Backend Engineering          292
IT & Security                252
Name: count, dtype: int64

# 7. Extracting Finalized Data

In [ ]:
merged_job_df.to_csv('job_listings.csv', index=False)

# 8. Saving Knowledge Base

### Loading Summarized Dataset

In [14]:
import pandas as pd

df = pd.read_csv("job_listings.csv")

### Cleaning Rules for Skills

In [15]:
# Skills appearing less than this in listings will be removed
MIN_FREQUENCY = 5

# Skills longer than this many words are sentences, not skills
MAX_WORDS = 4

# These keywords indicate benefits/requirements, not skills
GARBAGE_KEYWORDS = [
    'leave', 'bonus', 'pension', 'discount', 'insurance', 'salary',
    'benefit', 'holiday', 'degree', 'years of experience', 'assurance',
    'remote work', 'hybrid', 'full-time', 'part-time', '401k', 'referral',
    'ability to work', 'closure', 'healthcare', 'dental', 'vision'
]

# These appear in every category and are useless for gap detection
SOFT_SKILLS = {
    'communication', 'teamwork', 'collaboration', 'problem solving',
    'problemsolving', 'innovation', 'reliability', 'leadership',
    'time management', 'critical thinking', 'adaptability', 'creativity',
    'troubleshooting'  # too generic, add this
}

### Defining Validation Function

In [16]:
import re

def is_valid_skill(skill, freq):
    # Rule 1: must appear in at least 5 listings
    if freq < MIN_FREQUENCY:
        return False
    # Rule 2: more than 5 words = it's a sentence, not a skill
    if len(skill.split()) > MAX_WORDS:
        return False
    # Rule 3: contains benefit/requirement keywords
    if any(g in skill for g in GARBAGE_KEYWORDS):
        return False
    # Rule 4: generic soft skill, useless for gap detection
    if skill in SOFT_SKILLS:
        return False
    # Rule 5: starts with a number (e.g. "10+ years", "401k")
    if re.match(r'^\d', skill):
        return False
    return True


### Building the Knowledge Base

In [17]:
from collections import defaultdict

# Count how many times each skill appears per job category
knowledge_base = defaultdict(lambda: defaultdict(int))

for _, row in df.iterrows():
    category = row['job_title']
    skills = [s.strip().lower() for s in str(row['job_skills']).split(',')]
    for skill in skills:
        knowledge_base[category][skill] += 1

print("Raw knowledge base built.")
for cat, skills in knowledge_base.items():
    print(f"  {cat}: {len(skills)} raw skills")

Raw knowledge base built.
  Network Engineering: 8043 raw skills
  Software Engineering: 14626 raw skills
  Data Engineering: 3569 raw skills
  Data Science & Analytics: 5105 raw skills
  DevOps & Automation: 3835 raw skills
  IT & Security: 2481 raw skills
  Backend Engineering: 1117 raw skills


### Cleaning the Knowledge Base

In [18]:
# Apply all validation rules to filter out noise
cleaned_kb = {}

for category, skills in knowledge_base.items():
    cleaned_kb[category] = {
        skill: freq
        for skill, freq in skills.items()
        if is_valid_skill(skill, freq)
    }

for cat, skills in cleaned_kb.items():
    print(f"  {cat}: {len(skills)} skills remaining")

  Network Engineering: 739 skills remaining
  Software Engineering: 1375 skills remaining
  Data Engineering: 348 skills remaining
  Data Science & Analytics: 509 skills remaining
  DevOps & Automation: 201 skills remaining
  IT & Security: 231 skills remaining
  Backend Engineering: 233 skills remaining


### Top 10 skills per category

In [19]:
for cat, skills in cleaned_kb.items():
    top10 = sorted(skills.items(), key=lambda x: x[1], reverse=True)[:10]
    print(f"\n{cat}:")
    for skill, freq in top10:
        print(f"  {skill}: {freq}")


Network Engineering:
  network engineering: 635
  dns: 540
  ansible: 433
  network security: 415
  python: 413
  tcp/ip: 409
  network troubleshooting: 381
  scripting: 371
  powershell: 364
  bgp: 363

Software Engineering:
  java: 957
  python: 943
  aws: 808
  software engineering: 779
  javascript: 737
  sql: 735
  software development: 668
  docker: 649
  kubernetes: 605
  agile: 494

Data Engineering:
  python: 440
  sql: 341
  data engineering: 277
  spark: 239
  java: 234
  aws: 225
  scala: 208
  snowflake: 208
  data warehousing: 203
  kafka: 188

Data Science & Analytics:
  python: 470
  sql: 420
  machine learning: 344
  data analysis: 340
  data visualization: 283
  data science: 218
  r: 217
  tableau: 193
  statistics: 148
  tensorflow: 130

DevOps & Automation:
  python: 105
  automation: 104
  kubernetes: 93
  devops: 85
  terraform: 84
  docker: 83
  aws: 80
  jenkins: 75
  ansible: 75
  bash: 70

IT & Security:
  cybersecurity governance: 123
  fisma: 118
  nist 80

### Saving and Downloading the .pkl File

In [20]:
import pickle

with open('knowledge_base.pkl', 'wb') as f:
    pickle.dump(cleaned_kb, f)

print("Saved!")

# Download to your local machine
from google.colab import files
files.download('knowledge_base.pkl')

Saved!


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>